### ISSUE TITLE
Two bugs

### ISSUE THREAD

**Issue by LiaoMengqi** [GeeeekExplorer/nano-vllm#144]
1. bug for tensor parallelism

When using tensor parallelism, the `if self.num_completion_tokens == 0` is used to determine whether it's prefilling or decoding. However, this can lead to errors because of situations where preempted requests re-prefill. We solve this problem by adding a `prefilled` field to the Sequence. Postprocess sets this field to `True` and sets it to `False` when preemption occurs.

2. bug for prefilling and shared prefix

During prefilling, if the length is exactly a multiple of the block size and all blocks hit the cache, `seqlen_q = 0`, which will cause `flash_attn_varlen_func` to throw an error. We currently solve this problem by forcibly setting the hash of the last block to -1.

**Comment by fishAndShrimp** [comment]
Hi @LiaoMengqi ,

I also noticed nano-vLLM might have this potential bug while building my own LLM framework from scratch (it runs, but I'm still suffering through CUDA optimizations 🥲) and designing the radix/prefix tree state machine.

For this edge case, my engine adopted an "L-1" approach, which is actually very similar to how vLLM V1 handles full prompt hits. If the prompt is fully cached, my engine just subtracts 1 from seqlen (or num_computed_tokens) instead of discarding the entire last matched block.

This way, during the next forward pass, the exact same KV cache is harmlessly overwritten into the block, and we naturally get the logits to predict the next token. So the engine only needs to reprefill 1 token rather than recomputing the whole block!

Here is the L-1 implementation in my scheduler, in case you are interested:
https://github.com/fishAndShrimp/femto-vllm/blob/main/femtovllm/engine/scheduler/v3.py#L370

**Comment by fishAndShrimp** [comment]
Just a quick update: I went ahead and implemented this L-1 approach in PR #199. Feel free to test it out! 🚀
